### PR TITLE
KAFKA-12495: Improve rebalance allocation algorithm

### PR DESCRIPTION
**Pull request by C0urante**
Follow-up as discussed on the [open PR](https://github.com/apache/kafka/pull/12561) for KAFKA-12495, which ports some changes from https://github.com/apache/kafka/pull/12019 onto that PR branch.

These changes should fix some issues with load-balancing revocations and assignment of new and lost connectors and tasks. The primary differences are when load-balancing revocations take place (a number of currently-missed cases are added), when lost assignments are reassigned, and how load-balancing revocations are calculated.

One result is that a number of unnecessary follow-up rebalances are skipped since a balanced allocation can be achieved sooner, and potential regressions introduced in the upstream PR for KAFKA-12495 are avoided as a result.

Another result is that resizing connectors should no longer lead to unbalanced allocations across the cluster during rebalance.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
This is ready for review now, but I've marked it as a draft to signal that it should not be merged into the upstream branch yet. I'm hoping we can review this PR separately until it's approved to be merged into the trunk branch of the upstream repo, then merge it and the upstream KAFKA-12495 branch together into trunk all at once.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: nit: Could `deletedAndRevoked` be called `deletedAndToRevoke`? IIUC, that's the intent of this right ie these resources are `deleted` and would be `revoked`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Same comment here.. `duplicatedAndRevoked` could be renamed to `duplicatedAndToRevoke`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: Curious, is this the result of the new Load Balancing assignment that you have come up with instead of the floor/ceiling based mechanism that was there before?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:279]: `nextWorkerAssignment` basically signifies the workerAssignments for this round, right? It has `next` because it eventually gets assigned to `previous` after the rebalance round?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: i couldn't follow this comment..

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:714]: I maybe missing something here, but the goal looks like to have each worker have either <m or m + 1> resource. If it has less that min, doesn't it become a candidate to receive some resources from others which have higher. Let's say it has 3 and minToAllocate is 5 and there's some other worker having 7, then 2 can be reduced from the latter and assigned to this one. I see below the logic to remove extra resources from workers.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: i think this could be called `workersWithExtraLoad` to signify `How many workers are going to have to be allocated exactly one extra instance`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:724]: Just for my understanding, you mean that this worker has extra worker but we haven't reached the `allocatedExtras` cap yet, so this worker won't see any revocations right?

**Review by vamossagar12 — COMMENTED** [review]
@C0urante , Thanks for the follow up PR! I have a few comments mainly for my understanding and some naming related stuffs which I felt could change

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:248]: Chris, this is something I get confused a lot with and not a comment for this PR but in general. There are a lot of steps which happen before we get to actual revocations. Can we just add a comment to explain how we get to it? Like `configured - duplicate - deleted` . I know it's all in the code but becomes difficult to follow at times ( could be due to my limited working with this class as well). Let me know what you think about it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Good call--maybe for brevity's sake, we could do `deletedToRevoke` (and similarly below)?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: Yep, it is! As long as it's safe to allocate them (which it should be in this case), it's better to see more connectors and tasks allocated instead of less, since that means that we're decreasing downtime while they're being moved across workers, and moving more quickly towards a balanced, stable assignment.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:279]: That's correct, `nextWorkerAssignment` is the assignment that will be given out after this round completes. There's three kinds of assignment we care about during this round:
- The assignment that we sent out during the last round (`previousAssignment`)
- The assignment that everyone has reported in their metadata while joining the group during the current rebalance (`currentWorkerAssignment`)
- The assignment that we'll give out at the end of the current rebalance (`nextConnectorAssignments`)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: When we detect duplicates, we revoke them from every worker that's running them, then wait for the next round to reassign them to a single worker, and this all takes place outside this method. The line here just clarifies that we don't take that into account when performing load-balancing revocations, and instead pretend that any duplicates that we're revoking in this round are allowed to be assigned. This is fine since the very next round should result in those duplicates being assigned, at which point a balanced allocation should be achieved.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:714]: > If it has less that min, doesn't it become a candidate to receive some resources from others which have higher.

Exactly! The idea here (in the `performLoadBalancingRevocations` method) is just to do the revoking part of this logic. The other part--where we assign connectors and tasks to workers that may have fewer than the minimum--is handled later on, when we invoke `assignConnectors` and `assignTasks`. It may also take a follow-up rebalance for this to take place, since we have to wait for workers to stop running their revoked connectors and tasks before those can be assigned to different workers.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: I can see how `extrasToAllocate` is misleading, but I'm a little hesitant to use `workersWithExtraLoad` since it could be interpreted as "the number of workers that are running too many connectors/tasks".

Thoughts on `workersToAllocateExtra`? It's a little verbose but should be fairly clear.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:724]: Yep, exactly 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:248]: I've found this part of the code base difficult to read and reason about as well. I'm not sure what would be helpful with this comment in particular, though?

If there's something in this PR that would benefit from extra comments, I'm happy to add them here until things look good. For other parts, we should probably leave that for a cosmetic follow-up that reduces the diff and allows us to focus on readability (and possibly make other, more aggressive, refactoring).

**Review by C0urante — COMMENTED** [review]
Thanks @vamossagar12, this is ready for another round when you have time. I've also removed some now-unused methods, as a minor cleanup step.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: `deletedToRevoke` is good.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: `duplicatedAndToRevoke` is good.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: Makes sense. This is great! Just a follow up question on `we're decreasing downtime while they're being moved across workers`, what are the advantages of moving more tasks v/s less? Because eventually in this test it converged to a balanced load. PS: I am not against this change at all but just for my understanding.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: `workersToAllocateExtra` sounds good to me.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:714]: Ok. I think this is where the other comment you made about being able to assign more connectors/tasks in a round makes more sense. This could mean we could converge faster.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:248]: What I meant was how do we start from the assignments and reduce it to the set with which we finally operate. I look at the code again and it's definitely less verbose in this PR v/s what it is in trunk. I can't call out what exactly to comment, but something that would help people figure out faster of the steps performed. We can take it later as well, not a problem.

**Comment by vamossagar12** [comment]
> Thanks @vamossagar12, this is ready for another round when you have time. I've also removed some now-unused methods, as a minor cleanup step.

Thanks @C0urante . I made another pass.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: The idea is that, if we've revoked some tasks, the sooner we reassign them, the better. There are some inefficiencies in the current allocation algorithm that cause tasks that were revoked in a prior round to not be reassigned in a future round, even when it would be perfectly safe to do so. This PR helps address some of those inefficiencies.

Of course, you're correct that in both cases, things converge to an even load, so in most cases, this optimization will be relatively minor.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Does this make sense now? Any changes you'd like to see?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:248]: 👍 sounds good. A few things we might consider:
- Compute derived sets immediately before they're used, instead of in a single large block at the beginning of the method
- Reduce comment size by removing the copy+pasted "Derived set: The set of... is a derived set from the set..." snippet
- Remove the `WorkerLoad` class and try to consolidate its logic into the `ConnectorsAndTasks` class (and the various `Map<String, ConnectorsAndTasks>` instances that we use)

(Disclaimer: haven't put much thought into this, these could be valuable, or we could find out in practice that they either aren't that useful or even actively make things worse)

**Comment by C0urante** [comment]
Thanks @vamossagar12. Ready for the next round when you have time.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: Oh. I wasn't aware of this (maybe missed it) => `There are some inefficiencies in the current allocation algorithm that cause tasks that were revoked in a prior round to not be reassigned in a future round, even when it would be perfectly safe to do so`. I think in this case it makes sense to do what you are suggesting.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Ok thanks for the explanation. I think this comment makes sense now.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:248]: Yeah. I think these are not very important right now and we can have follow up JIRA tickets on this if needed to clean some things up.

**Comment by vamossagar12** [comment]
@C0urante , I looked at it again. The changes look good to me and can't think of new comments at this point. Thanks for the changes !

**Comment by C0urante** [comment]
Thanks @vamossagar12!

@showuon if you can spare a few cycles, mind taking a look? Once this is ready to go we should be able to finally merge a fix for [KAFKA-12495](https://issues.apache.org/jira/browse/KAFKA-12495)!

**Comment by showuon** [comment]
Sorry, quite busy these days. I'll try to review it next week.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:712]: I think we can add a comment above this for loop to mention we're going to calculate how many connectors/tasks to revoke from existing worker

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:226]: Nice!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:502]: Why changed to 0 delay now?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:1097]: Nice improvement

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java:526]: Nice fix!

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (showuon) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:271]: Just want to confirm if we don't have this patch, this test will fail in https://github.com/apache/kafka/pull/12561 , right?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: I don't have preference here, but we changed the variable name from `newSubmissions` to `created`, should we also update the comment?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: I don't understand this comment:
`Any workers that left the group within the scheduled rebalance delay permanently left the group`.
The word `permanently` seems to be confusing to me.

Comment (showuon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: TBH, the javadoc can be more clear if we remove all these 3 bullets, because I felt more confused after reading these 3 bullets. So, maybe change to:
`Revoke connectors and tasks from each worker in the cluster until no worker is running more than it would be`

WDYT?

**Review by showuon — COMMENTED** [review]
@C0urante , thanks for the improvement. Overall LGTM! Just some minor comments.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:712]: 👍 Done

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:502]: Due to optimizations in the allocation algorithm, we no longer perform a task-balancing revocation in the fourth round, which allows us to reset the delay.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:271]: That's correct 👍

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: 👍 Good call, done

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: Fine with simplifying this 👍 
I updated it to `Revoke connectors and tasks from each worker in the cluster until no worker is running more than it would be with a perfectly-balanced assignment.`; LMKWYT

**Comment by C0urante** [comment]
Thanks @showuon! Pushed a new commit; ready when you have time.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java
Comment (showuon) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:502]: I see.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java
Comment (showuon) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java]: LGTM

**Review by showuon — APPROVED** [review]
LGTM! Great improvement!

**Comment by vamossagar12** [comment]
Thanks for your time @showuon ! @C0urante , how do we want to take this further? Merge this into my branch and merge that or you want to create another PR against trunk and merge the 2 of them together? I guess it would be better to do the latter to highlight the great improvements you have made here (former might be simpler I reckon)

**Comment by showuon** [comment]
Merge it into this branch, and check the jenkins build status

**Comment by vamossagar12** [comment]
@showuon , Chris had a different idea I think => 
```
This is ready for review now, but I've marked it as a draft to signal that it should not be merged into the upstream branch yet. I'm hoping we can review this PR separately until it's approved to be merged into the trunk branch of the upstream repo, then merge it and the upstream KAFKA-12495 branch together into trunk all at once.
```

Would wait for his response.

**Comment by C0urante** [comment]
@vamossagar12 it's fine to merge this one now :)

**Review by vamossagar12 — APPROVED** [review]
LGTM

**Comment by vamossagar12** [comment]
@C0urante done! Thanks for your contribution on this one!

**Pull request merged**
